### ISSUE TITLE
Chunk identifier is too long and too ambiguous

### ISSUE THREAD

**Issue by korelstar** [x00/Consolidate-Vanilla-Plugin#3]
The method `_Chunk($Files, $Suffix)` clusters files and saves them into an array by using a generated array key. Currently, this key is generated using this code:
`$NiceKey = trim(Gdn_Format::Url(preg_replace('`[^a-z0-9]+`i','_',$Chunk)),'_').'_';`

This has two disadvantages:
1. It is too ambiguous: If I define two chunks using the patterns `/foo+bar/.*` and `/foo-bar/.*`, respective files will be consolidated into one single chunk, because both `+` and `-` are replaced by `_`.
2. It is too long: I have defined a base set of CSS files, which are definitely used on every page, so caching can be utilized in the best way. However, the pattern is quite long: `(themes/bootstrap/design/(style|custom)|resources/design/vanillicon|plugins/editor/design/editor|plugins/Tagging/design/(childtagslist|tag))\.css`. The resulting name for the consolidated file has a length of 194 characters! Such long file names can be problematic on certain file systems and waste bandwidth.

As as solution, I would propose to change the `$NiceKey` to a short checksum of the pattern. Optionally, the first `n` characters of the pattern could be appended to the pattern in the old way. E.g.:

`$NiceKey = md5($Chunk).'_';`

or

`$NiceKey = trim(Gdn_Format::Url(preg_replace('`[^a-z0-9]+`i','_',substr($Chunk, 0, 20))),'_').'_'.hash('crc32b', $Chunk).'_';`

**Comment by korelstar** [comment]
Here is another (better) solution:
`$NiceKey = substr($Suffix,1).$cid.'-';`
where `$cid` is the chunk number from the array -- hence the foreach-loop has to be changed:
`foreach($Chunks As $cid => $Chunk){`

**Comment by x00** [comment]
I think you have a point.

I think one hash is enough. 

The question is more over the usefulness of the identifier prefix. For reference it can be useful, but not essential as part of the url. 

Maybe a subsisting would work, which we could have a config setting for the length. The default should be a reasonable length.

**Comment by x00** [comment]
I think the `NiceKey` should not be the array index. So there is no chance of collision.

I favour renaming the `NiceKey` to `NicePrefix` which is the same method but with a maximum length set by config, and passing that with the files.

**Comment by korelstar** [comment]
You're right: one hash is enough. And yes: the current array index from `$Chunks` would be better as array index for `$ChunkedFiles` (btw. there is another bug in this code: `$ChunkedFile` should be `$ChunkedFiles`).

However, I really like to use the array index for naming the files. I'm using the suffix as prefix for convenience when looking into the cache directory or source code for debugging purposes. It's much easier than having the cleaned up PCRE patterns as file name, since these are not unique! And my solution generates shorter file names.

Example for the files in my forum:

```
css0-7fe8410bfbcf8e52f07e8fdc3a4ae905.css
js0-53f1cda64e14c6fcd171f0fc549e7ed5.js
js1-0db631085cfc23cd47e2cb7dac4e2173.js
js2-39ad07a1ae4cf34967c5712961bc0fc8.js
js3-84b4e73cc10b19e67dee9ba08bc26a0f.js
```

**Comment by x00** [comment]
The lack of uniqueness of the nice preffix isn't a big deal. It can be handled by making sure the content doesn't get overwritten via key, the hash itself is unique with a low chance of collision.  It is always taken on the serialised matching files, so that is clear as mud. 

We could encode more characters for clarity, however I would still rather have some indication of where the file came from albeit up to a reasonable length. 

IMO the array index prefix is not a good idea. It is something that exists within the context of the a request and you are caching the files across requests.  It unnecessary anyway. It is not what holds things to be unique. 

The nice prefix is there for ease of reference, however if you don't want to use it you could set the length to 0 and only the hash would appear.

**Comment by korelstar** [comment]
I think you misunderstood me (might be due to my bad Englisch). I want to use the array index of `$Chunks`. This is equivalent to the position of the matching pattern in the configuration file. This is static even across requests and gives a convenient indication of where the file came from, i.e. which pattern matched the file (in order to be even more convenient, you could increase the number by 1).

Example:

We have three pattern defined:
1. `js/library/jquery.*\.js`
2. `foo\+bar/.*`
3. `foo\-bar/.*`

There are four files included: `js/library/jquery.js`, `js/library/jquery.form.js`, `foo-bar/a.js` and `foo-bar/b.js`
My approach generates two consolidated files (please imagine the correct md5 sum; I've already applied the new idea to increase the number by 1 to be more clear):
1. `js1-123465789MD5.js` consisting of `js/library/jquery.js` and `js/library/jquery.form.js`
2. `js3-123456789MD5.js` consisting of `foo-bar/a.js` and `foo-bar/b.js`

Again: I would forget about using the cleaned patterns in file names, since in this example, a consolidated file `foo_bar-123456789MD5.js` doesn't give a clue, if it is generated by using pattern 1 or by pattern 2!

To be even clearer, I would make a pull request. However, there is already an unprocessed pull request (#2) concerning this code area and I do not want to produce merge conflicts.

**Comment by x00** [comment]
I not worried about similar patterns getting confused that is quite an edge case for normal use. Or if it is due to advanced matching, again it is not really a big deal. 

Also it not good convention to name your folders that way anyway, especially with `+` in it on a server. 

I'm a favour of keeping the descriptive part intact, as it gives some clue for simple matching. I will just make sure it is not too long, and you could simply set the length to `0` if if you don't want it. 

The the pattern index Ok we can put that to help with any ambiguity.

**Comment by korelstar** [comment]
> I will just make sure it is not too long, and you could simply set the length to 0 if if you don't want it.
> The the pattern index Ok we can put that to help with any ambiguity.

Sounds good to me. :-)
